Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs

```diff
@@ -0,0 +1,327 @@
+        }
+
+        private void QueueCommand(QueuedNpmCommandInfo info)
+        {
+            lock (this.queuelock)
+            {
+                if (this.commandQueue.Contains(info)
+                    || info.Equals(this.currentCommand))
+                {
+                    return;
+                }
+                this.commandQueue.Enqueue(info);
+                Monitor.PulseAll(this.queuelock);
+            }
+        }
+
+        public void QueueCommand(string arguments)
+        {
+            QueueCommand(new QueuedNpmCommandInfo(arguments));
+        }
+
+        private async void Execute(QueuedNpmCommandInfo info)
+        {
+            this.IsExecutingCommand = true;
+            INpmCommander cmdr = null;
+            try
+            {
+                lock (this.queuelock)
+                {
+                    cmdr = this.npmController.CreateNpmCommander();
+
+                    this.commander = cmdr;
+                }
+
+                await cmdr.ExecuteNpmCommandAsync(info.Arguments);
+            }
+            finally
+            {
+                lock (this.queuelock)
+                {
+                    this.commander = null;
+                }
+                this.IsExecutingCommand = false;
+            }
+        }
+
+        private void Run()
+        {
+            var count = 0;
+            // We want the thread to continue running queued commands before
+            // exiting so the user can close the install window without having to wait
+            // for commands to complete.
+            while (!this.isDisposed || count > 0)
+            {
+                lock (this.queuelock)
+                {
+                    while ((this.commandQueue.Count == 0 && !this.isDisposed)
+                        || this.npmController == null
+                        || this.IsExecutingCommand)
+                    {
+                        Monitor.Wait(this.queuelock);
+                    }
+
+                    if (this.commandQueue.Count > 0)
+                    {
+                        this.currentCommand = this.commandQueue.Dequeue();
+                    }
+                    else
+                    {
+                        this.currentCommand = null;
+                    }
+                    count = this.commandQueue.Count;
+                }
+
+                if (null != this.currentCommand)
+                {
+                    Execute(this.currentCommand);
+                }
+            }
+        }
+
+        public async Task<IEnumerable<IPackage>> GetCatalogPackagesAsync(string filterText)
+        {
+            var relativeUri = string.Format("/-/v1/search?text={0}", filterText);
+            var searchUri = new Uri(defaultRegistryUri, relativeUri);
+
+            var request = WebRequest.Create(searchUri);
+            using (var response = await request.GetResponseAsync())
+            {
+                var reader = new StreamReader(response.GetResponseStream());
```

Comment:
Add a comment showing the expected format of the JSON so people can follow this later